incipyt/project.py

```diff
@@ -8,134 +9,133 @@
 
     .. code-block::
 
-        var_first = environ["VARIABLE_NAME"]
-        # Prompt a message for confirmation of the value for `VARIABLE_NAME`
-        var_second = environ["VARIABLE_NAME"]
-        # As `VARIABLE_NAME` as been confirmed, don't ask anything
+        EnvMetadata("VARIABLE_NAME", required=True)
 
-    Moreover, even if a specific variable hasn't been set, it can be asked for
-    it anyway and it will be confirmed immediately. A variable can also be set
-    and confirmed at the same time:
+    When there is no `default` suggestion, if no value is given when prompt asks for it
+    `None` is saved in `environ` and all :class:`incipyt._internal.templates.Formattable`
+    values will be purge from template dictionary before commiting. When `True`, an empty
+    string is saved in `environ`. Note that if the `default` suggestion is not empty,
+    `required` has no effect as the prompt result will never be an empty string.
+    """
 
-    .. code-block::
+    name: str
+    default: str = ""
+    required: bool = False
+
+    def __post_init__(self):
+        if self.default is None:
+            self.default = ""
+        if self.name not in metadata:
+            metadata[self.name] = self
+        else:
+            if self.default:
+                metadata[self.name].default = self.default
+            metadata[self.name].required |= self.required
 
-        environ["VARIABLE_NAME"] = EnvValue("something", confirmed=True)
 
-    Note that if a variable has been set, it cannot be set again, except if
-    explicitly specified:
+class _Environ(collections.UserDict):
+    """Manage environ variables using for substitutions in patterns.
+
+    A :class:`Environ` object `environ` is a dictionnary.
 
     .. code-block::
 
-        environ["VARIABLE_NAME"] = EnvValue("new_value", update=True)
+        environ["VARIABLE_NAME"] = "something"
 
-    All iterative methods and in operator consider only `confirmed` variables.
+    If a specific variable hasn't been set, it can be asked for it anyway and
+    it will be add to the `environ` immediately.
     """
 
     def clear(self):
-        self._confirmed.clear()
         self.data.clear()
-        self.data.update(os.environ.copy())
 
-        self.data.setdefault("PYTHON_CMD", sys.executable)
-        self._confirmed.add("PYTHON_CMD")
-        self.data.setdefault("YEAR", date.today().year)
-        self._confirmed.add("YEAR")
+        self.data["PYTHON_CMD"] = sys.executable
+        self.data["YEAR"] = date.today().year
+
+        metadata.clear()
+
+        for var_name, var_value in os.environ.items():
+            if var_name.startswith("INCIPYT__"):
+                if var_name.removeprefix("INCIPYT__") in ("CHECK_BUILD",):
+                    self.data[var_name.removeprefix("INCIPYT__")] = strtobool(var_value)
+                else:
+                    self.data[var_name.removeprefix("INCIPYT__")] = var_value
+            EnvMetadata(var_name, default=var_value)
```

Comment:
Dans la plupart des outils que je connais, le prefix n'utilise qu'un simple underscore. Je serais d'avis qu'on fasse pareil, juste `INCIPYT_` au lieu d'`INCIPYT__`.

Pour la partie `if  ... in ("CHECK_BUILD",)` l'idée de pouvoir plutôt faire `self.data[var_name] = metadata[var_name].type(var_value)` à la place me semble plus futur-proof. cf mon commentaire plus haut

Quand je lis ceci j'ai quand même une question importante niveau `EnvMetadata` et valeur par défaut. Ici on voit assez clairement que niveau Metadata on n'est pas vraiment au top... On init bien une metadata mais genre si un jour on veut rajouter un attribut `help` bah ce sera quand même pas génial.

En vrai de vrai, ici ça crie pour la ChainMap. Enfin ça crie que le dictionnaire `os.environ` devrait être utilisé tel-quel et être placé dans une ChainMap. Plus prioritaire que les véritables valeurs par défaut.

Dans l'état actuel, si je démarre incipyt avec une variable d'environnement `AUTHOR_NAME="julien"` (sans le prefix `INCIPYT_`), je ne sais pas si ce sera cette valeur ou bien celle déterminée par le `git config user.name` qui sera la `EnvMetadata.default`, je pense que ne pas pouvoir intuitivement répondre à cette question montre un mauvais design.

Reply:
Je suis d'accord pour le `ChainMap` je n'étais pas allé dans cette direction pour le garder pour plus tard. Mais, ça vaut peut-être le coup de le mettre en place maintenant.